From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hpfxd <me@hpfxd.com>
Date: Tue, 26 Oct 2021 21:33:07 -0400
Subject: [PATCH] Backport server tick events from Paper

These events were added by Paper in the late 1.13.2 era, and are used by plugins such as anti-cheats and performance monitors to tell when a tick has started and ended.

diff --git a/src/main/java/com/destroystokyo/paper/event/server/ServerTickEndEvent.java b/src/main/java/com/destroystokyo/paper/event/server/ServerTickEndEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..e16c093a408eda45cc806f53e38af3565e36d17c
--- /dev/null
+++ b/src/main/java/com/destroystokyo/paper/event/server/ServerTickEndEvent.java
@@ -0,0 +1,56 @@
+package com.destroystokyo.paper.event.server;
+
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+
+/**
+ * Called when the server has finished ticking the main loop
+ */
+public class ServerTickEndEvent extends Event {
+
+    private static final HandlerList handlers = new HandlerList();
+    private final int tickNumber;
+    private final double tickDuration;
+    private final long timeEnd;
+
+    public ServerTickEndEvent(int tickNumber, double tickDuration, long timeRemaining) {
+        this.tickNumber = tickNumber;
+        this.tickDuration = tickDuration;
+        this.timeEnd = System.nanoTime() + timeRemaining;
+    }
+
+    /**
+     * @return What tick this was since start (first tick = 1)
+     */
+    public int getTickNumber() {
+        return tickNumber;
+    }
+
+    /**
+     * @return Time in milliseconds of how long this tick took
+     */
+    public double getTickDuration() {
+        return tickDuration;
+    }
+
+    /**
+     * Amount of nanoseconds remaining before the next tick should start.
+     *
+     * If this value is negative, then that means the server has exceeded the tick time limit and TPS has been lost.
+     *
+     * Method will continously return the updated time remaining value. (return value is not static)
+     *
+     * @return Amount of nanoseconds remaining before the next tick should start
+     */
+    public long getTimeRemaining() {
+        return this.timeEnd - System.nanoTime();
+    }
+
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
diff --git a/src/main/java/com/destroystokyo/paper/event/server/ServerTickStartEvent.java b/src/main/java/com/destroystokyo/paper/event/server/ServerTickStartEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..6aa0d1febde1498b75dcd858035ed2dba9b1a3e5
--- /dev/null
+++ b/src/main/java/com/destroystokyo/paper/event/server/ServerTickStartEvent.java
@@ -0,0 +1,29 @@
+package com.destroystokyo.paper.event.server;
+
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+
+public class ServerTickStartEvent extends Event {
+
+    private static final HandlerList handlers = new HandlerList();
+    private final int tickNumber;
+
+    public ServerTickStartEvent(int tickNumber) {
+        this.tickNumber = tickNumber;
+    }
+
+    /**
+     * @return What tick this is going be since start (first tick = 1)
+     */
+    public int getTickNumber() {
+        return tickNumber;
+    }
+
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
